src/python/py/models/builder.py

```diff
@@ -1839,6 +1859,12 @@ def make_model(self, input_path):
             elif module.__class__.__name__.endswith("DecoderLayer") and self.layer_id < self.num_layers:
                 # Each decoder layer of model
                 print(f"Reading decoder layer {self.layer_id}")
+
+                # Some model stores rotary_emb at model class instead of attention objects.
+                if hasattr(model, "model"):
+                    if hasattr(model.model, "rotary_emb"):
```

Comment:
After re-reading the [modeling code](https://github.com/huggingface/transformers/blob/3f06f95ebe617b192251ef756518690f5bc7ff76/src/transformers/models/nemotron/modeling_nemotron.py#L85), I think we can remove this altogether and remove the need to override the `make_attention` method in the Nemotron class. The cos/sin caches for the rotary embeddings in Nemotron are calculated in the same way as LLaMA.

https://github.com/microsoft/onnxruntime-genai/blob/921e0f47eb078c53ecd9d0dbd5c8e7e1930ebc9c/src/python/py/models/builder.py#L995-L1000

Since there isn't a `cos_cached` attribute, the cos/sin caches should get calculated correctly for the user.

Reply:
Not passing a valid rotary embedding leads to crashes in make_attention(), it expects "attention.rotary_emb" to be valid object. So, overriding make_attention feels necessary to patch rotary_emb as suggested in review earlier.

`                   # Make RotaryEmbedding nodes
        cos_cache_name, sin_cache_name = "", ""
        if self.attention_attrs["use_rotemb_in_attn"]:
            cos_cache_name, sin_cache_name = self.make_rotary_embedding_caches(attention.rotary_emb)
        else:
            q_rotary_name = f"/model/layers.{layer_id}/attn/q_rotary/RotaryEmbedding"
            self.make_rotary_embedding(attention.rotary_emb, q_rotary_name, root_input=q_input_to_attention, position_ids=kwargs.get("position_ids", "position_ids"))
            q_input_to_attention = f"{q_rotary_name}/output_0"
            k_rotary_name = f"/model/layers.{layer_id}/attn/k_rotary/RotaryEmbedding"
            self.make_rotary_embedding(attention.rotary_emb, k_rotary_name, root_input=k_input_to_attention, position_ids=kwargs.get("position_ids", "position_ids"))
            k_input_to_attention = f"{k_rotary_name}/output_0"

`

Reply:
I think we can make a dummy rotary embedding attribute for `attention.rotary_emb` instead.

```python
def make_attention(self, layer_id, attention, root_input, **kwargs):
    attention.rotary_emb = type("RotaryEmbedding", (object,), {'content':{}})()
    return super().make_attention(layer_id, attention, root_input, **kwargs)
```
[This PR](https://github.com/microsoft/onnxruntime-genai/pull/921) has an example of how this is done.

https://github.com/microsoft/onnxruntime-genai/blob/25636b434bfb93564be304b3218931a68fb6ac9f/src/python/py/models/builder.py#L2646-L2651

With this dummy attribute, the logic for `model_rotary_emb` can be removed and `attention.rotary_emb` is still defined.

Reply:
Sure, that makes sense. Fixed in the latest code now.